Optimize HTTP2 HPack huffman decoding

rokonec opened the pull request:
Optimized to use 8 bits lookup tables tree with  result of about 0.35 CPU utilization as oppose to former version.
Decoding table is lazy generated as ushort[].

dotnet/runtime#1506

After this changes goes through review and receive final approve, I will create corresponding pull request to aspnetcore including these changes and new benchmark similar to [existing hpack benchmark](https://github.com/dotnet/aspnetcore/blob/master/src/Servers/Kestrel/perf/Kestrel.Performance/HPackDecoderBenchmark.cs)

rokonec committed:
Optimize HTTP2 HPack huffman decoding

Optimized to use 8 bits lookup tables tree with  result of about 0.35 CPU utilization as oppose to former version.
Decoding table is lazy generated as ushort[].

runtime#1506

rokonec committed:
Fix nullability

stephentoub commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -307,128 +400,147 @@ public static (uint encoded, int bitLength) Encode(int data)
+            if (s_decodingTree == null)
+            {
+                Volatile.Write(ref s_decodingTree, GenerateDecodingLookupTree());
+            }
```
Consider storing s_decodingTree into a local and then using that local instead of the static field on all future accesses in this method.

rokonec replied:
done

stephentoub commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -307,128 +400,147 @@ public static (uint encoded, int bitLength) Encode(int data)
+                    lookupIndex = (byte)(acc >> (bitsInAcc - 8));
 
-                dst[j++] = (byte)ch;
+                    var lookupValue = s_decodingTree[(lookupTableIndex << 8) + lookupIndex];
```
Nit: var => whatever type this is

rokonec replied:
done

stephentoub commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -307,128 +400,147 @@ public static (uint encoded, int bitLength) Encode(int data)
-            for (int i = 0; i < _decodingTable.Length && _decodingTable[i].codeLength <= validBits; i++)
-            {
-                (int codeLength, int[] codes) = _decodingTable[i];
+                var lookupValue = s_decodingTree[(lookupTableIndex << 8) + lookupIndex];
```
Ditto

rokonec replied:
dome

stephentoub commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -2,11 +2,14 @@
     {
+        private static ushort[]? s_decodingTree;
+
         // TODO: this can be constructed from _decodingTable
```
_decodingTable is being deleted

rokonec replied:
Shall I just remove comment or add github link to former code?

stephentoub replied:
Is the comment still relevant to s_decodingTree, i.e. could/should the encoding table also be constructed dynamically?  If yes, you can update the comment (or address it).  If not, you can delete it.

rokonec replied:
I think that _encodingTable shall not be generated. We need either _encodingTable or _decodingTable just to be able to generate  _encodingTable from it.
Both are about the same size:
![image](https://user-images.githubusercontent.com/25249058/96896303-a40a1300-148d-11eb-866b-45bc9446b3a7.png)
but _decodingTable has more heap allocations.
I will replace the comment by refernce to RFC from which is this table build

rokonec replied:
done

rokonec committed:
Review comments #1

- removing var where unclear
- remove outdated comments
- using local variable as oppose to static

rokonec committed:
Add #nulable enable

rokonec commented:
Closed/reopened to pickup fixes from master - worked :-)

ManickaP commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -269,36 +273,126 @@ internal class Huffman
             return _encodingTable[data];
         }
 
+        private static ushort[] GenerateDecodingLookupTree()
```
Trying to understand this whole function has been quite a feat since it's quite a lot of bit twiddling. In the end, the generated table is static and its content is always the same. So my question is, do we need this method?

Could we just put static data in and some nice comment above explaining them?

rokonec replied:
I have discussed it with @scalablecory > https://github.com/dotnet/runtime/issues/1506#issuecomment-707347243
If I understand it correctly the recommended tradeoff is to not enlarge size of assembly binary, but rather lazy generate table.  It makes sense to me, as System.Net.Http will very in wast majority cases used without HTTP 2 functionality.

ManickaP reviewed (commented):
At the end of `GenerateDecodingLookupTree`, will continue on Monday.

rokonec commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -269,36 +273,126 @@ internal class Huffman
+            // Create traverse path for all 0..256 octets, 256 is EOS, see: http://httpwg.org/specs/rfc7541.html#rfc.section.5.2
+            for (int octet = 0; octet <= 256; octet++)
+            {
+                (uint code, int bitsLeft) = Encode(octet);
```
I also frown on it a little when I was writing it. I think this will be more readable:
```C#
(uint code, int bitLength) = Encode(octet);
int bitsLeft = bitlenght;
```

rokonec commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -269,36 +273,126 @@ internal class Huffman
+            // -----------------------------------------------------------------
+            //  15  14  13  12  11  10   9   8   7   6   5   4   3   2   1   0
+            // +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+            // | 0 |   next_lookup_table_index |            not_used           |
```
This is very good point.
I have changed it.

rokonec commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -269,36 +273,126 @@ internal class Huffman
+
+            // Because HPack uses static huffman code defined in RFC https://httpwg.org/specs/rfc7541.html#huffman.code
+            // it is guaranteed that for this huffman code generated decoding lookup tree MUST consist of 15 lookup tables
+            var lut = new ushort[15 * 256];
```
Renamed

rokonec committed:
Review changes, ManickaP  #1
- renaming LUT
- swapping leaf flag meaning (MSB)
- typos

ManickaP commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -307,128 +402,148 @@ public static (uint encoded, int bitLength) Encode(int data)
+                if (lookupTableIndex == 0)
+                {
+                    // Check if all remaining bits are ones.
+                    uint ones = uint.MaxValue >> (64 - bitsInAcc);
```
Is this correct? Type is `uint` (32 bits) but you're rotating by 57-63 positions. I [works](https://dotnetfiddle.net/9kXObi), don't get me wrong, but it's just thanks to this C# feature: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/bitwise-and-shift-operators#shift-count-of-the-shift-operators.

rokonec replied:
Good catch. It is lucky bug.
Fixed by replacing 64 => 32.
BTW, that C# feature https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/bitwise-and-shift-operators#shift-count-of-the-shift-operators is quite controversial to me as it differes from C/C++ behavior. the fact that `((ulong)0x1 >> 64) == 0x1` is not very intuitive.

ManickaP commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -269,36 +273,127 @@ internal class Huffman
+
+            // Because HPack uses static huffman code defined in RFC https://httpwg.org/specs/rfc7541.html#huffman.code
+            // it is guaranteed that for this huffman code generated decoding lookup tree MUST consist of exactly 15 lookup tables
+            var decodingTree = new ushort[15 * 256];
```
Actually you can just use 2-dimensional, rectangular array: `new ushort[15, 256]`. And then indexed access of `decodingTree[lookupTableIndex, (indexInLookupTable | suffix)]`. It will be represented in memory as single dim array and compiler/runtime will handle the offset calculations for you.

rokonec replied:
Tempting, I will verify performance of this and if it will be same, and I expect it will, I will change it. 
To be continued...

gfoidl replied:
> 2-dimensional, rectangular array

The runtime can't handle these as good as sz-arrays (see e.g. https://github.com/dotnet/runtime/issues/35293, https://github.com/dotnet/runtime/issues/5481) and indexing is mostly method-based as opposed to direct (JIT) support for the sz-arrays.

So I doubt the perf with the rectangular array can be on par.

rokonec replied:
Benchmark:
|      Method | DataIndex |         Mean |     Error |    StdDev |         Op/s | Ratio |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------ |---------- |-------------:|----------:|----------:|-------------:|------:|-------:|------:|------:|----------:|
|     Current |        -1 |    575.93 ns |  4.116 ns |  3.648 ns |  1,736,315.4 |  1.00 |      - |     - |     - |      64 B |
|      Tree2d |        -1 |    598.66 ns |  6.873 ns |  6.093 ns |  1,670,391.6 |  1.04 |      - |     - |     - |      64 B |

It is slightly slower. I believe it is because my custom index computing using fast 'x<<8' is faster than x*n when n is not known at compile time thus had to be read from array header.

On the other hand, it is slightly more readable with ushort[,]. But since array index computation is used only at about 5 places, I'd recommend to keep it as is.

What do you think?

gfoidl replied:
I'd leave it as is -- same as you said.

Another (minor) point is that the current implementation (sz-array) produces less JITed code, which is preferable (in most cases).
For readability current version is readable -- I even see no need to extract the index-computation to a (local) function as this is just overhead (for the JIT to inline this function).

rokonec replied:
Hmm. As for index computation I though compiler would inline it, wouldn't it?
I can inline it, it would not harm readability much. @gfoidl Shall I?

gfoidl replied:
> for index computation I though compiler would inline it,

Yeah, it will inline it. It's just a shift and add, so pretty small and straight-forward. As it's so "mini" I'd leave it manually inlined. Also the logic isn't very complicated.

I would extract it to a separate method if unsafe indexing is used, so that Debug.Asserts can be used to verify the access is within bounds. But this is not the case here. Just leave it as is.

ManickaP commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -307,128 +402,148 @@ public static (uint encoded, int bitLength) Encode(int data)
+            // Although bellow code could be inserted in above do..while,
+            // processing of last byte has to handle several corner cases was decided, for performance reasons,
+            // to not have in main decoding loop.
+            while (bitsInAcc > 0)
```
Can this go for more than 1 loop?

I'm not questioning the correctness here, just making sure I understand it correctly.

rokonec replied:
Yes it can. Example:
- suppose we last 7 bits left
- 2 bits could form end of currently decoding symbol (1st run)
- 5 bits would be next symbol (2nd run)

I have added unit test test vectors to specifically address this situation.
If one would put break; at then end of this loop, couple of unit tests would fail.

ManickaP replied:
> 2 bits could form end of currently decoding symbol (1st run)

This was what I didn't realize could happen. Thanks for explaining.

ManickaP commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -307,128 +402,148 @@ public static (uint encoded, int bitLength) Encode(int data)
+            int lookupTableIndex = 0;
+            int lookupIndex;
+
+            ulong acc = 0;
```
Is there a reason for 64bit `acc`? Is it perf? Or am I missing something fundamental about the algorithm? :smile:

rokonec replied:
It would work for 16 bits and 32 bits acc as well. By having longest possible accumulator for most common OS (64 bits) shifting operations to get next 8 bits is very well compiled into Amd64 machine code by JIT. Same would apply for 32 bits acc, but the bigger the acc the less transition between loading and decoding data would occur.
Those would be line 431 & 480.
However, if I would read it by bytes I could safe lines 434 & 439 !
I will do some perf test for 16 bit accumulator with eliminating loop at line 439
I'll be back :-)

rokonec replied:
It turned out, after running benchmarks, that reading just next byte into UInt16 is actually faster, with ration of 0.93, and, to me, more readable. Additionally it eliminates suboptimal performance of dealing with ulong in 32 bit OS.
Have changed it.

|      Method | DataIndex |         Mean |     Error |    StdDev |         Op/s | Ratio |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------ |---------- |-------------:|----------:|----------:|-------------:|------:|-------:|------:|------:|----------:|
|     Current |        -1 |    575.93 ns |  4.116 ns |  3.648 ns |  1,736,315.4 |  1.00 |      - |     - |     - |      64 B |
| Acc16bits |        -1 |    536.60 ns |  3.289 ns |  3.076 ns |  1,863,588.2 |  0.93 |      - |     - |     - |      64 B |

ManickaP reviewed (commented):
In general LGTM. Some little comments and questions. Thank you!

BTW, this should go through ASP.NET folks before merging. Do you have a person of contact or should I invoke someone?

Tratcher commented:
> BTW, this should go through ASP.NET folks before merging. Do you have a person of contact or should I invoke someone?

That's me 😁. 

> Decoding table is lazy generated as ushort[].

You're trading disk size for startup time. How long in ms does it take generate the table?

Tratcher commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -307,128 +402,148 @@ public static (uint encoded, int bitLength) Encode(int data)
 
+            if (s_decodingTree == null)
+            {
+                Volatile.Write(ref s_decodingTree, GenerateDecodingLookupTree());
```
The Volatile.Write only guarantees you've flushed the result to memory, it doesn't guarantee anybody reads it, you'd need a Volatile.Read for that which you don't want happening on every request. You could end up running this initialization code on every core at startup. What's the cost?

rokonec replied:
I have discussed this tradeoff with Cory. I am aware about possibility of multiple initialization with current implementation.

Bellow is a benchmark.

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
AMD Ryzen 5 3600, 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=6.0.100-alpha.1.20472.11
  [Host]     : .NET Core 6.0.0 (CoreCLR 6.0.20.47901, CoreFX 6.0.20.47901), X64 RyuJIT
  Job-LKQRKP : .NET Core 6.0.0 (CoreCLR 6.0.20.46807, CoreFX 6.0.20.46807), X64 RyuJIT

Server=True  Toolchain=.NET Core 5.0  RunStrategy=Throughput

|               Method |     Mean |     Error |    StdDev |      Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------- |---------:|----------:|----------:|----------:|-------:|------:|------:|----------:|
| GenerateDecodingTree | 5.296 us | 0.0441 us | 0.0391 us | 188,805.9 | 0.0687 |     - |     - |   7.52 KB |

Please let me know if 5us is acceptable or if you recommend some further optimization in this area.

Tratcher replied:
Ok, that's pretty cheep. Thanks for checking.

stephentoub replied:
> you'd need a Volatile.Read for that which you don't want happening on every request

Technically you need a volatile read anyway.

rokonec committed:
Review changes, ManickaP #1
- fix shift count computation
- change acc size to 16 bits and append to it by byte

rokonec committed:
Update src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs

Co-authored-by: Marie Píchová <11718369+ManickaP@users.noreply.github.com>

rokonec committed:
Update src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs

Co-authored-by: Marie Píchová <11718369+ManickaP@users.noreply.github.com>

rokonec committed:
Update src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs

Co-authored-by: Marie Píchová <11718369+ManickaP@users.noreply.github.com>

scalablecory commented on src/libraries/Common/src/System/Security/Cryptography/Asn1/AlgorithmIdentifierAsn.xml.cs:
```diff
@@ -1,4 +1,4 @@
-// Licensed to the .NET Foundation under one or more agreements.
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
It looks like your editor added a BOM to this and other files. Lets revert this.

rokonec replied:
I have reverted it all.

scalablecory commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -307,128 +402,145 @@ public static (uint encoded, int bitLength) Encode(int data)
 
-                int mask = int.MinValue >> (codeLength - 1);
-                long masked = (data & mask) >> (32 - codeLength);
+                    if (j >= dst.Length)
```
nit: looks like `j` should never be > `dst.Length`. can make this a little clearer by changing comparison.

```suggestion
                    if (j == dst.Length)
```

scalablecory commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -307,128 +402,145 @@ public static (uint encoded, int bitLength) Encode(int data)
+                        // +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+                        // | 0 |     number_of_used_bits   |              octet            |
+                        // +---+---------------------------+-------------------------------+
+                        if (j >= dst.Length)
```
```suggestion
                        if (j == dst.Length)
```

rokonec committed:
Revert partially "Review changes, ManickaP #2"

This reverts commit a09b4d30c6f5662a9d92e1334a247bb33502078d, partially, all but Huffman.cs

rokonec committed:
Merge branch 'rokonec/1506-optimize-huffman-decoding-using-lut' of https://github.com/rokonec/runtime into rokonec/1506-optimize-huffman-decoding-using-lut

scalablecory commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -269,36 +273,127 @@ internal class Huffman
+                        // +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+                        // | 1 |   next_lookup_table_index |            not_used           |
+                        // +---+---------------------------+-------------------------------+
+                        if (lookupValue == default(ushort))
```
```suggestion
                        if (lookupValue == 0)
```

scalablecory reviewed (commented):
lgtm once outstanding comments are addressed. I'm not an expert in Huffman -- lets please make sure tests are exercising everything.

rokonec committed:
Update src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs

Co-authored-by: Cory Nelson <phrosty@gmail.com>

rokonec committed:
Apply suggestions from code review

Co-authored-by: Cory Nelson <phrosty@gmail.com>

rokonec commented:
As for testing, there were really good unit tests extend created.
To easy my mind, I have created temporary tests, in my test branch, which created random bytes test vector from size 1..7 bytes and compare result of new and former decoding. Then I run it for about one hour.
Since the code is not very complex above plus unit tests gave me sufficient confidence that the code is not defective.

gfoidl commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -307,128 +402,144 @@ public static (uint encoded, int bitLength) Encode(int data)
+            int lookupTableIndex = 0;
+            int lookupIndex;
+
+            ushort acc = 0;
```
Is there an advantage of using `ushort`? The cpu needs to (zero) extend (`movzx`) here.
Using `uint` avoids some of these.

For the array `ushort` is fine, but for the locals (~ "registers") just use `(u)int`.

rokonec replied:
done

gfoidl commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -307,128 +402,144 @@ public static (uint encoded, int bitLength) Encode(int data)
-                    Array.Resize(ref dstArray, dst.Length * 2);
-                    dst = dstArray;
-                }
+                    lookupIndex = (byte)(acc >> (bitsInAcc - 8));
```
Make `uint lookupIndex` then it's better for registers (no partial registers use):
```suggestion
                    lookupIndex = acc >> (bitsInAcc - 8);
```
Or is the cast to `byte` needed to mask something out?

rokonec replied:
yes, `(byte)` has role of `& 0xff`

gfoidl commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -307,128 +402,144 @@ public static (uint encoded, int bitLength) Encode(int data)
+                    lookupIndex = (byte)(acc >> (bitsInAcc - 8));
 
-                dst[j++] = (byte)ch;
+                    ushort lookupValue = decodingTree[(lookupTableIndex << 8) + lookupIndex];
```
```suggestion
                    uint lookupValue = decodingTree[(lookupTableIndex << 8) + lookupIndex];
```
Same reason, avoid partial registers, as they need to be zero-extended (`movzx`).

rokonec replied:
done

gfoidl commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -307,128 +402,144 @@ public static (uint encoded, int bitLength) Encode(int data)
-            for (int i = 0; i < _decodingTable.Length && _decodingTable[i].codeLength <= validBits; i++)
-            {
-                (int codeLength, int[] codes) = _decodingTable[i];
+                ushort lookupValue = decodingTree[(lookupTableIndex << 8) + lookupIndex];
```
```suggestion
                uint lookupValue = decodingTree[(lookupTableIndex << 8) + lookupIndex];
```

rokonec replied:
done

gfoidl commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -307,128 +402,144 @@ public static (uint encoded, int bitLength) Encode(int data)
+            {
+                // Finished in middle of traversing - no valid symbol could be decoded
+                // or too long EOS padding (7 bits plus). See: http://httpwg.org/specs/rfc7541.html#rfc.section.5.2
+                throw new HuffmanDecodingException(SR.net_http_hpack_huffman_decode_failed);
```
This exception is 4 times in this method, maybe move this to a throw-helper?

I.e.
```c#
private static void ThrowHuffmanDecodingException() => throw new HuffmanDecodingException(SR.net_http_hpack_huffman_decode_failed);
```
and call this method instead of throwing directly.
This should move the exception-throwing into a cold part (machine code wise), and reduce the code size in `Decode`-method.

rokonec replied:
I have benchmarked recommended changes and did not measured significant perf difference.
I'll keep it as is for code maintainability.,

scalablecory commented:
/azp list

gfoidl commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -307,128 +402,144 @@ public static (uint encoded, int bitLength) Encode(int data)
             int i = 0;
             int j = 0;
-            int lastDecodedBits = 0;
             while (i < src.Length)
```
From the assert in L414 we know that there will be at least one element, so could change this to da `do-while`-loop, thus eliminating one length-check before loop entry.

rokonec replied:
L414 asserts dst not src. Expecting src has at least one byte would break contract of previous implementation.

gfoidl replied:
Oh, sorry...

scalablecory commented:
/azp run

rokonec committed:
Fix trailing ws

rokonec committed:
Merge branch 'rokonec/1506-optimize-huffman-decoding-using-lut' of https://github.com/rokonec/runtime into rokonec/1506-optimize-huffman-decoding-using-lut

gfoidl commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -307,128 +402,144 @@ public static (uint encoded, int bitLength) Encode(int data)
+                        // +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+                        // | 0 |     number_of_used_bits   |              octet            |
+                        // +---+---------------------------+-------------------------------+
+                        if (j == dst.Length)
+                        {
+                            Array.Resize(ref dstArray, dst.Length * 2);
+                            dst = dstArray;
+                        }
+                        dst[j++] = (byte)lookupValue;
```
```suggestion
                        if ((uint)j < (uint)dst.Length)
                        {
                            dst[j++] = (byte)lookupValue;
                        }
                        else
                        {
                            Array.Resize(ref dstArray, dst.Length * 2);
                            dst = dstArray;
                        }
```
This pattern has two points:
* the `uint` casts allow the JIT to eliminate the bound-checks on `dst[j]`
* the first branch is the more likely, and cpu's branch predictor will choose the fallthrough-path when no history for that branch is available

You can use this on other places as well.

gfoidl replied:
Ahhh, I've forgotten the write in the case of resizing.
Maybe this cold-portion of code should be used to a non-inlineable method, and there the resive and write after the resize can happen.

The idea is to keep the hot code as small as possible.

rokonec replied:
Why should hot code be as small as possible in this particular case?

rokonec replied:
I have benchmarked recommended changes and did not measured significant perf difference.
I'll keep it as is for code maintainability.,

gfoidl replied:
The case `j == dst.Length` is unlikely, at least it will happen less often than the other branch (`j < dst.Length`). In the case resize is needed, we have to pay the cost of resizing anyway, so we could move this part out into a cold region. This will make the hot region smaller, which is a loop (L439-L476) in this case.

Smaller loops means it's more likely that the cpu instructions can be read from an internal buffer / µop-cache or the (instruction) cache(line) instead of farther away memory hierarchies (which are slower) -- in short: the cpu's frontend has less work to do, which improves overall execution speed (also for other parts of the app or other processes). 

Side note: if the loop is small enough (size varies depending on cpu, but say 64 µops) then the cpu has a special loop-stream-detector (LSD). This means the cpu recognizes the loop and is able to use more efficient scheduling of instructions (roughly spoken). But the loops here are too big for this...

gfoidl replied:
As hint: in a micro-benchmark the cpu-caches are hot, which means the data and instruction are very likely in the cache (due to the tight benchmark iterations). This may not be the case in real world usage, where the cpu may evict these items from it's caches (by other data and instructions). This also falls into the "general rule" of smaller code is better (increased locality).

gfoidl reviewed (commented):
Some low-level micro-optimization hints.

rokonec commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -1,13 +1,17 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+#nullable enable
```
This file will be shared with aspnetcore which do not have enabled globally. 
Using `#nullable enable` is consistent with other files in related areas, for example: https://github.com/dotnet/runtime/blob/5f96636b63bea718bcfa1355e17cb36693ff8bca/src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackDecoder.cs#L4

gfoidl replied:
Thanks!

rokonec committed:
Review comments, gfoidl #1
- change local variables from ushort to int

rokonec commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -307,128 +402,144 @@ public static (uint encoded, int bitLength) Encode(int data)
 
+            if (s_decodingTree == null)
+            {
+                Volatile.Write(ref s_decodingTree, GenerateDecodingLookupTree());
```
I have benchmarked recommended changes and did not measured significant perf difference.
I'll keep it as is for code maintainability.,

stephentoub replied:
I assume you benchmarked on x86/x64?  Volatile.Read/Write evaporate completely from the codegen because of the strong memory model on such architectures.

halter73 commented:
Do you have benchmark results yet @rokonec?

rokonec commented:
> Do you have benchmark results yet @rokonec?

@halter73 What exact benchmark result do you mean?

geoffkizer commented:
How many entries are there in the decodingTable?

geoffkizer commented on src/libraries/Common/tests/Tests/System/Net/aspnetcore/Http2/HuffmanDecodingTests.cs:
```diff
@@ -354,29 +356,6 @@ public void HuffmanEncode(int code, uint expectedEncoded, int expectedBitLength)
         }
 
-        [Theory]
-        [MemberData(nameof(HuffmanData))]
```
Why were these removed?

rokonec replied:
Method Huffman.DecodeValue has been deleted during rewriting decode. It was, if I understand it correctly, internal only for purpose of unit testing this particular method and its contract was not used anywhere else, apart from previous Decode and these tests.

When I was deleting it, I was thinking of rewriting those test to use Huffman.Decode method however I realized that all test cases that this deleted tests have covered are also covered in existing HuffmanDecoding_ValidEncoding_Succeeds test which test cases are including all single element array of chars with value 0..255.

geoffkizer replied:
Ok, as long as we have coverage for all the cases that are removed, it's fine to get rid of these test cases.

rokonec commented:
> How many entries are there in the decodingTable?

@geoffkizer decodingTable is ushort[15*256] => 7.5 KB

geoffkizer commented:
Can we get the benchmark code added to the PR as well? This is useful for measuring future improvements.

geoffkizer commented:
BTW, I still suspect that the "jump table" algorithm discussed on the linked issue would be both faster and simpler.

rokonec commented:
> BTW, I still suspect that the "jump table" algorithm discussed on the linked issue would be both faster and simpler.

@geoffkizer The implementation in this PR is based on this comment https://github.com/dotnet/runtime/issues/1506#issuecomment-416000494 

If I understand it correctly "jump table" algorithm is based on N-ary Tree. Before implementing this PR I have created benchmarks which tests effectivity and decoding table size for different N. https://github.com/dotnet/runtime/issues/1506#issuecomment-705860700

8bits has been agreed as best tradeoff between speed and memory. This makes it 256-ary tree and that is the algorytm in this PR.

rokonec committed:
Optimize initialization and memory footprint of s_encodingTable
Based on comment: https://github.com/dotnet/corefx/pull/32124/files#r215467614

rokonec commented:
Based on [comment on another PR](https://github.com/dotnet/corefx/pull/32124/files#r215467614) I have decided to optimize memory footprint and initialization of _encodingTable.
As a result it should save about 5.5K IL bytes, 9.5K of IL Asm and related CPU time required for static constructor JIT and execution.

geoffkizer commented:
> Based on comment on another PR I have decided to optimize memory footprint and initialization of _encodingTable.

I don't think we should do this as part of this PR. We should keep this PR focused on decoding. If changing the encoding table is worthwhile, we should do it as a separate PR.

BTW it's not obvious to me why your change to encoding table initialization would save so much; can you explain?

rokonec commented:
> > Based on comment on another PR I have decided to optimize memory footprint and initialization of _encodingTable.
> 
> I don't think we should do this as part of this PR. We should keep this PR focused on decoding. If changing the encoding table is worthwhile, we should do it as a separate PR.

I can see your point, I've looked at it as overall optimization of Huffman.cs file for both memory and performance. Since Jan Kotas recommended such optimization in a former related PR, so I've thought I would opportunistically refactor this as well.
But if you feel like it better be in another PR, I have no problem to create new issue and related PR. Shall I?
 
> BTW it's not obvious to me why your change to encoding table initialization would save so much; can you explain?

It is because the way C# compiler initialize arrays. Arrays of primitive types and enums are compiled into metadata token as `bytearray` and during initialization copied into new allocated array. For more complex arrays, for each element compiler generates code  `array2[0] = new ValueTuple<int, int>(1, 2);` For 257 elements it is quite a lot of IL code which needs to be JITed and generates a lot of Asm machine code => 9.5K+ in our case.

[SharpLab example](https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgYGYAE9MGFMDeqmJmADgE5gBuAhsAKaZgB2wA2gLqYD6AZrQDOwTAF5MLBgHdmbLsVJEUpFZgQAaOOvTqALOoCs6gGzqFJAL4BuFefJU6jTAApWwdbOABKLr0EQAexlxSRlXNg83H047JVVSZwQABk0krzNleJd0FN00jKyXAxTjNLtrLNQLIA===)

For example this class:
```C#
public class C {
    private int[] _fast = new int[]
    {
        1,2,3,4,5,6,
    };    
    private (int, int)[] _slow = new (int, int)[]
    {
        (10,20),
        (30,40),
        (50,60)
    };       
}
```

would be compiled as:
```
    .method public hidebysig specialname rtspecialname instance void .ctor () cil managed 
    {
        .maxstack 6
       
       // initialization of  _fast
        IL_0000: ldarg.0
        IL_0001: ldc.i4.6
        IL_0002: newarr [System.Private.CoreLib]System.Int32
        IL_0007: dup
        IL_0008: ldtoken field valuetype '<PrivateImplementationDetails>'/'__StaticArrayInitTypeSize=24' '<PrivateImplementationDetails>'::C4E70AB31EF6C8908F896CAD1C6BC75F7FA65E27
        IL_000d: call void [System.Private.CoreLib]System.Runtime.CompilerServices.RuntimeHelpers::InitializeArray(class [System.Private.CoreLib]System.Array, valuetype [System.Private.CoreLib]System.RuntimeFieldHandle)
        IL_0012: stfld int32[] C::_fast

        // initialization of  _slow
        IL_0017: ldarg.0
        IL_0018: ldc.i4.3
        IL_0019: newarr valuetype [System.Private.CoreLib]System.ValueTuple`2<int32, int32>
        IL_001e: dup
        IL_001f: ldc.i4.0
        IL_0020: ldc.i4.s 10
        IL_0022: ldc.i4.s 20
        IL_0024: newobj instance void valuetype [System.Private.CoreLib]System.ValueTuple`2<int32, int32>::.ctor(!0, !1)
        IL_0029: stelem valuetype [System.Private.CoreLib]System.ValueTuple`2<int32, int32>
        IL_002e: dup
        IL_002f: ldc.i4.1
        IL_0030: ldc.i4.s 30
        IL_0032: ldc.i4.s 40
        IL_0034: newobj instance void valuetype [System.Private.CoreLib]System.ValueTuple`2<int32, int32>::.ctor(!0, !1)
        IL_0039: stelem valuetype [System.Private.CoreLib]System.ValueTuple`2<int32, int32>
        IL_003e: dup
        IL_003f: ldc.i4.2
        IL_0040: ldc.i4.s 50
        IL_0042: ldc.i4.s 60
        IL_0044: newobj instance void valuetype [System.Private.CoreLib]System.ValueTuple`2<int32, int32>::.ctor(!0, !1)
        IL_0049: stelem valuetype [System.Private.CoreLib]System.ValueTuple`2<int32, int32>
        IL_004e: stfld valuetype [System.Private.CoreLib]System.ValueTuple`2<int32, int32>[] C::_slow
        IL_0053: ldarg.0
        IL_0054: call instance void [System.Private.CoreLib]System.Object::.ctor()
        IL_0059: ret
    } // end of method C::.ctor
...
    // Fields
    .field assembly static initonly valuetype '<PrivateImplementationDetails>'/'__StaticArrayInitTypeSize=24' C4E70AB31EF6C8908F896CAD1C6BC75F7FA65E27 at I_0000283C
    .data cil I_0000283C = bytearray (
        01 00 00 00 02 00 00 00 03 00 00 00 04 00 00 00
        05 00 00 00 06 00 00 00
    )
```

gfoidl commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -12,270 +12,532 @@ internal class Huffman
+            0b11111111_11111111_11111111_11111100,
+        };
+
+        private static readonly byte[] s_encodingTableOfLenght = new byte[257]
```
```suggestion
        private static ReadOnlySpan<byte> s_encodingTableOfLenght => new byte[257]
```
To use C# compiler optimization for one-byte-sized static data. See [this blog](https://vcsjones.dev/2019/02/01/csharp-readonly-span-bytes-static/) for further infos.

rokonec replied:
This has been moved to another PR

geoffkizer commented:
Thanks for the explanation, makes sense.

The downside of splitting the encoder table into two separate tables is that it messes up cache locality. 

Regardless, I do think this would be better done as a separate PR.

rokonec committed:
Revert "Optimize initialization and memory footprint of s_encodingTable"

This reverts commit 36d6cf49eef82cf225c247fd05c09512a2d82bd4.

geoffkizer commented:
btw, @stephentoub suggested that we can fix initialization cost while avoiding breaking locality by having two statically initialized tables (as in your change) but then combine them into one at runtime. Seems like that would address all the concerns here.

geoffkizer commented:
The "jump table" algorithm (don't know why we started calling it that, but let's go with it) is a different algorithm than your lookup tree algorithm. I don't think you've tried it out (unless I missed it?).

Looking back at the issue, I don't think there's a good clear description of it there; I'll put one together when I have a minute. 

I can't say for sure that it's better than the lookup table approach, but it seems worth investigating.

geoffkizer commented:
The "jump table" algorithm is an optimization of the classic, naive algorithm for decoding bit-by-bit using a binary tree.

In the classic algorithm, you start at the root of the tree and process one bit at a time. If the bit is 0, you walk to the left; if it's 1 you walk to the right. When you hit a leaf, you emit the associated character and start over at the root.

Of course you wouldn't literally represent this as a tree in memory. Instead you'd assign each internal node in the tree a number, from 0 (root) to 256. Then you'd have a lookup table that takes (node#, bit) and maps it to (isLeaf, emittedChar, nextNode). Note that emittedChar is only valid when isLeaf is true, and nextNode is always 0 for leaf nodes. 

The interesting observation here is that you don't need to walk the tree just one bit at a time. You can build a lookup table that takes multiple bits at a time. The best choice here seems to be 4, as it's easy to get 4 bits at a time from a byte, and it keeps the lookup table from getting too huge.

So, your lookup table is now (node#, 4bits) -> (isCharEmitted, emittedChar, nextNode). Note that isCharEmitted means that a character was emitted in walking any of those 4 bits -- and since all chars need at least 5 bits to encode, we know only one can be emitted per 4 bits. And note that nextNode now is independent of whether a character was emitted or not -- we could have emitted a character after the first bit, but nextNode will give the node that we ended up on after returning to the root and walking the next three bits.

So the inner loop here basically looks like:

```
  bits = getnext4bits();
  (isCharEmitted, emittedChar, currentNode) = lookuptable[currentNode, bits];
  if (isCharEmitted)
    output[index++] = emittedChar;
```

There's a little bit more complexity to handle invalid sequences (EOS) and the end of the encoding.

rokonec commented:
I have had carefully read your description and I do not see difference between "jump table" and my algorithm.

Here is what my loop body does:
```
bits = getnext8bits();
(isCharEmitted, emittedChar, unusedBits, nextNode) = lookuptable[currentNode, bits];
if (isCharEmitted)
  output[index++] = emittedChar;
  keepBits(unusedBits); // not all of 8 bits are used by emitted char, keep them for next code
else
  currentNode=nextNode;
```
Because we consume 8 bits it could well be that particular node could have enough nodes for multiple emitted characters. For example if we would have 9-bit-code,5bit-code. I tried to write data structure which would allow multiple chars to be emitted but not only the decoding table was 50% bigger (in bytes size) it was slower in benchmark due to added complexity in consuming emitted char. So in such situation we simple keep 7 bits from last lookup/jump, add 1 bit from source, reset to root and start over.
We have chosen 8 bits as it render best performance, about twice as fast as 4 bit, with acceptable size of decoding tree.

rokonec commented:
>Can we get the benchmark code added to the PR as well? This is useful for measuring future improvements.

@geoffkizer After this PR gets approved, I will create corresponding PR to aspnetcore, [as described here](https://github.com/dotnet/aspnetcore/blob/master/src/Shared/runtime/ReadMe.SharedCode.md) also creating new benchmark close to [existing hpack benchmark](https://github.com/dotnet/aspnetcore/blob/master/src/Servers/Kestrel/perf/Kestrel.Performance/HPackDecoderBenchmark.cs)  [here](https://github.com/dotnet/aspnetcore/blob/master/src/Shared/runtime/ReadMe.SharedCode.md)

Expected Benchmark looks like: https://gist.github.com/rokonec/4f63d464fd9a359add75056789e0f9e1

geoffkizer commented:
@Tratcher can you review this as well?

Tratcher reviewed (approved):
I don't know huffman, but I don't see anything alarming and I'm content with the test coverage we have here.

I've added @halter73 in case he wants to take a look.

rokonec committed:
Merge remote-tracking branch 'upstream/master' into rokonec/1506-optimize-huffman-decoding-using-lut

rokonec committed:
Merge branch 'master' of https://github.com/dotnet/runtime into rokonec/1506-optimize-huffman-decoding-using-lut

rokonec committed:
Merge branch 'master' of https://github.com/dotnet/runtime into rokonec/1506-optimize-huffman-decoding-using-lut

rokonec committed:
Added benchmark

ManickaP commented on src/libraries/System.Net.Http/tests/PerformanceTests/HPackHuffmanBenchmark/HPackHuffmanBenchmark.csproj:
```diff
@@ -0,0 +1,19 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Compile Include="$(CommonPath)System\Net\Http\aspnetcore\Http2\Hpack\Huffman.cs" Link="Common\System\Net\Http\aspnetcore\Http2\Hpack\Huffman.cs" />
```
```suggestion
    <Compile Include="$(CommonPath)System\Net\Http\aspnetcore\Http2\Hpack\Huffman.cs"
             Link="Common\System\Net\Http\aspnetcore\Http2\Hpack\Huffman.cs" />
```

And other linked files similarly.
And yes, I know, VS puts it on the same line, but for readability of csproj it was decided to split the lines (see #1266 and follow up discussion in the linked PR).

rokonec replied:
Done

ManickaP commented on src/libraries/System.Net.Http/tests/PerformanceTests/HPackHuffmanBenchmark/HPackHuffmanBenchmark.csproj:
```diff
@@ -0,0 +1,19 @@
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net5.0</TargetFramework>
+    <Nullable>enable</Nullable>
```
```suggestion
    <Nullable>enable</Nullable>    
    <LangVersion>9.0</LangVersion>
```
Known issue for BDN 12.1, see: dotnet/BenchmarkDotNet#1452

rokonec replied:
Done

ManickaP commented on src/libraries/System.Net.Http/tests/PerformanceTests/HPackHuffmanBenchmark/HPackHuffmanBenchmark.csproj:
```diff
@@ -0,0 +1,19 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
```
```suggestion
    <StringResourcesPath>../../../src/Resources/Strings.resx</StringResourcesPath>
    <OutputType>Exe</OutputType>
```
Instead of Fakes/SR.cs. The same way it's solved in FunctionalTests:
https://github.com/dotnet/runtime/blob/3bcd31d5a1567fd022f41c375bb3953db94cc5dd/src/libraries/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj#L3-L3

rokonec replied:
works as magic :-)

ManickaP reviewed (approved):
Just a few technicalities around the benchmark project, otherwise LGTM.
Thanks for this perf improvement!

ManickaP commented:
And please also update .gitignore because I'm seeing  `System.Net.Http/tests/PerformanceTests/HPackHuffmanBenchmark/BenchmarkDotNet.Artifacts` in un-commited changes.

stephentoub commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -307,128 +402,143 @@ public static (uint encoded, int bitLength) Encode(int data)
             Span<byte> dst = dstArray;
             Debug.Assert(dst != null && dst.Length > 0);
 
+            if (s_decodingTree == null)
+            {
+                Volatile.Write(ref s_decodingTree, GenerateDecodingLookupTree());
+            }
+            ushort[] decodingTree = s_decodingTree;
```
Technically this should be more like:
```suggestion
            ushort[] decodingTree = Volatile.Read(ref s_decodingTree);
            if (decodingTree is null)
            {
                decodingTree = GenerateDecodingLookupTree();
                Volatile.Write(ref s_decodingTree, decodingTree);
            }
```

That said, why is this not just:
```C#
private static readonly ushort[]? s_decodingTree = GenerateDecodingLookupTree();
```
?  Isn't it the case that if this type is being used, this s_decodingTree is going to be needed?

rokonec replied:
This is good point, In future, this class might handle encoding as well, but even in such cases, the situation that only encoding is used and decoding is not will be probably rare. As GenerateDecodingLookupTree is just ~5us we might not overoptimize it here. I will change it to `private static readonly ushort[] s_decodingTree = GenerateDecodingLookupTree();`

stephentoub commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -1,13 +1,17 @@
 
 namespace System.Net.Http.HPack
 {
     internal class Huffman
```
```suggestion
    internal static class Huffman
```

rokonec replied:
done

rokonec committed:
Review changes - stephentoub #2

rokonec committed:
Review changes - manickap #3

rokonec committed:
Fix, reorder static field for correct initialization order

stephentoub commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs:
```diff
@@ -269,36 +271,129 @@ internal class Huffman
-            (28, new[] { 2, 3, 4, 5, 6, 7, 8, 11, 12, 14, 15, 16, 17, 18, 19, 20, 21, 23, 24, 25, 26, 27, 28, 29, 30, 31, 127, 220, 249 }),
-            (30, new[] { 10, 13, 22, 256 })
-        };
+        private static ushort[] s_decodingTree = GenerateDecodingLookupTree();
```
```suggestion
        private static readonly ushort[] s_decodingTree = GenerateDecodingLookupTree();
```

rokonec committed:
Review changes stephentoub #4

Co-authored-by: Stephen Toub <stoub@microsoft.com>
